src/codegate/ca/codegate_ca.py

```diff
@@ -41,194 +108,179 @@ def __init__(self):
-                    content_commitment=False,
-                    key_encipherment=True,
-                    data_encipherment=False,
-                    key_agreement=False,
-                    key_cert_sign=True,  # This is a CA
-                    crl_sign=True,
-                    encipher_only=False,
-                    decipher_only=False,
-                ),
-                critical=True,
-            )
-
-            self._ca_cert = builder.sign(
-                private_key=self._ca_key,
-                algorithm=hashes.SHA256(),
-            )
+        self._ca_cert_expiry = None
+        self._ca_last_load_time = None
 
-            # Save CA certificate and key
-            if not os.path.exists(Config.get_config().certs_dir):
-                logger.debug(f"Creating directory: {Config.get_config().certs_dir}")
-                os.makedirs(Config.get_config().certs_dir)
+        # Use strong references for caching certificates
+        self._cert_cache: Dict[str, CachedCertificate] = {}
+        # Use a separate cache for SSL contexts
+        self._context_cache: Dict[str, Tuple[ssl.SSLContext, datetime]] = {}
 
-            with open(ca_cert, "wb") as f:
-                logger.debug(f"Saving CA certificate to {ca_cert}")
-                f.write(self._ca_cert.public_bytes(serialization.Encoding.PEM))
+        # Add a lock for thread-safe cache operations
+        self._cache_lock = threading.Lock()
+        CertificateAuthority._instance = self
 
-            with open(ca_key, "wb") as f:
-                logger.debug(f"Saving CA key to {ca_key}")
-                f.write(
-                    self._ca_key.private_bytes(
-                        encoding=serialization.Encoding.PEM,
-                        format=serialization.PrivateFormat.PKCS8,
-                        encryption_algorithm=serialization.NoEncryption(),
+        # Load existing certificates into cache
+        self._load_existing_certificates()
+
+    def _load_existing_certificates(self) -> None:
+        """Load existing certificates from disk into the cache."""
+        logger.debug("Loading existing certificates from disk into cache")
+        certs_dir = Config.get_config().certs_dir
+
+        if not os.path.exists(certs_dir):
+            logger.debug(f"Certificates directory {certs_dir} does not exist")
+            return
+
+        # First load the CA certificate to verify signatures
+        try:
+            ca_cert_path = self.get_cert_path(Config.get_config().ca_cert)
+            logger.debug(f"Loading CA certificate for verification: {ca_cert_path}")
+            with open(ca_cert_path, "rb") as f:
+                ca_cert = x509.load_pem_x509_certificate(f.read(), default_backend())
+        except Exception as e:
+            logger.error(f"Failed to load CA certificate for verification: {e}")
+            return
+
+        # Get current time for expiry checks
+        current_time = datetime.now(timezone.utc)
```

Comment:
I had to look up in the docs if datetime.now is equivalent to datetime.utcnow and based on my understanding it is - just wanted to double check during review if you think it's OK to use both in the code (not a blocker)